lib/phoenix_live_view/channel.ex

```diff
@@ -338,41 +353,63 @@ defmodule Phoenix.LiveView.Channel do
 
   defp gather_keys(nil, acc), do: acc
 
-  defp handle_changed(state, %Socket{} = new_socket, ref, pending_live_patch \\ nil) do
+  defp handle_changed(
+         state,
+         %Socket{} = new_socket,
+         {from, arity},
+         ref,
+         pending_live_patch \\ nil
+       ) do
     new_state = %{state | socket: new_socket}
+    root_pid = new_socket.root_pid
 
     case maybe_changed(new_state) do
       :diff ->
         {diff, new_state} = render_diff(new_state, new_socket)
 
         {:ok, :diff,
          new_state
-         |> push_live_patch(pending_live_patch)
+         |> push_live_patch(pending_live_patch, {from, arity})
          |> push_render(diff, ref)}
 
       :noop ->
         {:ok, :noop,
          new_state
-         |> push_live_patch(pending_live_patch)
+         |> push_live_patch(pending_live_patch, {from, arity})
          |> push_noop(ref)}
 
       {:redirect, %{to: to} = opts} ->
-        new_state = push_redirect(new_state, opts, ref)
-        send(new_state.transport_pid, {:socket_close, self(), {:redirect, to}})
-        {:stop, {:shutdown, {:redirect, to}}, new_state}
+        new_state
+        |> push_redirect(opts, {from, arity}, ref)
+        |> stop_shutdown_redirect(to)
 
       {:live, :redirect, %{to: to} = opts} ->
-        new_state = push_live_redirect(new_state, opts, ref)
-        send(new_state.transport_pid, {:socket_close, self(), {:redirect, to}})
-        {:stop, {:shutdown, {:redirect, to}}, new_state}
-
-      {:live, {params, action}, %{to: _to, kind: _kind} = opts} ->
         new_state
-        |> drop_redirect()
-        |> sync_handle_params_with_live_redirect(params, action, opts, ref)
+        |> push_live_redirect(opts, {from, arity}, ref)
+        |> stop_shutdown_redirect(to)
+
+      {:live, {params, action}, %{to: _to, kind: _kind} = opts} = patch ->
+        if root_pid == self() do
```

Comment:
Consider maybe moving this conditional to a guard to remove the nesting:

    {:live, {params, action}, %{to: _to, kind: _kind} = opts} when root_pid == self() ->